promql/engine.go

```diff
@@ -2871,6 +2870,17 @@ func (ev *evaluator) aggregation(e *parser.AggregateExpr, q float64, inputMatrix
 				// point in copying the histogram in that case.
 			} else {
 				group.hasFloat = true
+				if !group.incrementalMean {
+					newV, newC := kahanSumInc(f, group.floatValue, group.floatKahanC)
+					if !math.IsInf(newV, 0) {
+						group.floatValue, group.floatKahanC = newV, newC
+						break
+					}
+					// Handle overflow by reverting to incremental calculation of the mean value.
+					group.incrementalMean = true
+					group.floatMean = group.floatValue / (group.groupCount - 1)
+					group.floatKahanC /= group.groupCount - 1
```

Comment:
What sort of situation makes this useful?  We've overflowed a float64, but dividing by groupCount brings us back within range?

Reply:
We _would_ have overflowed, but instead of actually doing it, we revert to incremental mean calculation, whose whole purpose is to avoid overflowing by "dividing early" (with some tricks). This is what we have done so far before this PR, even in (the vast majority of) cases that don't overflow.

I have tried to make the comment a bit more explanatory.

Reply:
Right, I mean what sort of data will bring us to this case.
Perhaps 1,000 series each of whose value is 2e305 ?
And I am hinting that it may not be a very common case.

Reply:
Yes, I would assume it's very rare in practice.
And that's one of the motivations behind this change.
Prior to this change, we always did incremental mean calculation, which is expensive and less accurate.
With this change, we do "normal" mean calculation, backed by Kahan summation, and only switch to incremental mean calculation in the rare cases that we need it.